File: packages/android/engine/src/main/kotlin/me/sospedra/blasphem/Artifacts.kt
package me.sospedra.blasphem

import android.content.res.AssetManager
import java.io.File
import java.io.FileNotFoundException
import java.io.IOException

/** The two files a locale ships: its pack and its language-detection slice. */
internal enum class ArtifactKind(private val extension: String, private val artifactPrefix: String) {
    PACK("pack", "blasphem-pack-"),
    DETECT("detect", "blasphem-detect-"),
    ;

    fun file(code: String): String = "$code.$extension"

    /** The Maven artifact that carries `<code>.<extension>`, such as `me.sospedra.blasphem:blasphem-pack-es`. */
    fun artifact(code: String): String = "me.sospedra.blasphem:$artifactPrefix$code"
}

internal typealias ReadArtifact = (code: String, kind: ArtifactKind) -> ByteArray

/** Reads `assets/blasphem/<code>.<kind>`, the path every `blasphem-pack-*` and `blasphem-detect-*` AAR merges into the app. */
internal fun assetReader(assets: AssetManager): ReadArtifact = { code, kind ->
    val name = "blasphem/${kind.file(code)}"
    try {
        assets.open(name).use { it.readBytes() }
    } catch (missing: FileNotFoundException) {
        throw BlasphemException(BlasphemException.Code.LOCALE_MISSING, "add ${kind.artifact(code)}")
    } catch (failure: IOException) {
        throw BlasphemException(BlasphemException.Code.FETCH_FAILED, "assets/$name: ${failure.message}")
    }
}

/** Reads `<code>.<kind>` from a folder. Tests and command-line hosts use it. */
internal fun directoryReader(directory: File): ReadArtifact = { code, kind ->
    val file = directory.resolve(kind.file(code))
    if (!file.isFile) {
        throw BlasphemException(BlasphemException.Code.LOCALE_MISSING, "${kind.file(code)} is not in ${directory.path}")
    }
    try {
        file.readBytes()
    } catch (failure: IOException) {
        throw BlasphemException(BlasphemException.Code.FETCH_FAILED, "${file.path}: ${failure.message}")
    }
}

File: packages/swift/Sources/Blasphem/Artifacts.swift
import Foundation

/// The two files a locale ships: its pack and its language-detection slice.
enum ArtifactKind {
    case pack
    case detect

    var fileExtension: String {
        switch self {
        case .pack: return "pack"
        case .detect: return "detect"
        }
    }

    /// The SwiftPM product that carries `<code>.<extension>`, such as `BlasphemPackES`.
    func product(_ code: String) -> String {
        switch self {
        case .pack: return "BlasphemPack\(code.uppercased())"
        case .detect: return "BlasphemDetect\(code.uppercased())"
        }
    }
}

/// The bytes of one artifact, from `packsDirectory` when set, otherwise from the
/// SwiftPM resource bundle `Blasphem_<product>.bundle` the app carries.
func readArtifact(code: String, kind: ArtifactKind, packsDirectory: URL?) throws -> Data {
    let file = "\(code).\(kind.fileExtension)"
    if let packsDirectory {
        let url = packsDirectory.appendingPathComponent(file)
        return try readFile(url, missing: "\(file) is not in \(packsDirectory.path)")
    }
    let product = kind.product(code)
    guard let url = resourceURL(product: product, code: code, fileExtension: kind.fileExtension) else {
        throw BlasphemError(code: .localeMissing, message: "add the product \(product) to the target")
    }
    return try readFile(url, missing: "add the product \(product) to the target")
}

/// The search the generated `Bundle.module` accessor performs, so it holds for
/// apps, app extensions, tests, and executables.
private func resourceURL(product: String, code: String, fileExtension: String) -> URL? {
    let bundleName = "Blasphem_\(product)"
    let candidates = [
        Bundle.main.url(forResource: bundleName, withExtension: "bundle"),
        Bundle(for: Judge.self).resourceURL?.appendingPathComponent("\(bundleName).bundle"),
        Bundle.main.bundleURL.appendingPathComponent("\(bundleName).bundle"),
    ]
    let bundle = candidates.lazy.compactMap { $0 }.compactMap(Bundle.init(url:)).first
    return bundle?.url(forResource: code, withExtension: fileExtension)
}

private func readFile(_ url: URL, missing: String) throws -> Data {
    guard FileManager.default.fileExists(atPath: url.path) else {
        throw BlasphemError(code: .localeMissing, message: missing)
    }
    do {
        return try Data(contentsOf: url)
    } catch {
        throw BlasphemError(code: .fetchFailed, message: "\(url.lastPathComponent): \(error.localizedDescription)")
    }
}
